Preview Sites: Display full date in the tooltip over relative site expiration time

ivan-ottinger opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Resolves https://github.com/Automattic/dotcom-forge/issues/10423.

## Proposed Changes

- add full date format to the tooltip displayed over the preview site expiration time
- widen columns to add more space for the relative date values

| Before | After |
|--------|--------|
| ![Markup on 2025-02-12 at 14:03:34](https://github.com/user-attachments/assets/274da1de-5397-480a-beb2-612ee51c8f28) | ![Markup on 2025-02-12 at 13:58:01](https://github.com/user-attachments/assets/4b846081-0e4b-49b7-af94-24513d3f36c2) | 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Check out the PR branch and build the app with `STUDIO_QUICK_DEPLOYS=true npm start`.
2. If you don't have any preview sites, create a couple of them.
3. When you hover over values in the `Expires` column, a tooltip with full expiration date should display (as can be seen in the screenshot above).
4. Columns should have more space than before.
5. You can test with different locales.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

ivan-ottinger committed:
Update dates

ivan-ottinger committed:
Use `min` instead of `m` for minutes

ivan-ottinger committed:
Fix tests

ivan-ottinger committed:
Remove unused `__` import

ivan-ottinger commented on src/lib/date.ts:
```diff
@@ -24,3 +26,27 @@ export const formatDistance: FormatDistanceFn = ( token, count ) => {
+export const formatDistanceShort: FormatDistanceFn = ( token, count ) => {
+	let stringToFormat = '';
+	switch ( token ) {
+		case 'xDays':
+			stringToFormat = _n( '%d d', '%d d', count );
+			break;
+		case 'xHours':
+			stringToFormat = _n( '%d h', '%d h', count );
+			break;
+		case 'xMinutes':
+			stringToFormat = _n( '%d min', '%d min', count );
+			break;
```
I left pluralization on purpose here - in case some languages need it.

ivan-ottinger replied:
This is not relevant anymore. → Closing this thread.

fredrikekelund commented on src/hooks/use-format-localized-timestamps.tsx:
```diff
@@ -1,32 +1,36 @@
+			if ( timestamp === null ) {
+				return '';
+			}
 
-		const duration = intervalToDuration( {
-			start: new Date( timestamp ),
-			end: new Date(),
-		} );
+			const duration = intervalToDuration( {
+				start: new Date( timestamp ),
+				end: new Date(),
+			} );
 
-		return formatTimeDistance( duration );
-	}, [] );
+			return formatTimeDistance( duration, format );
+		},
+		[]
+	);
```
Would it work to use `formatDistanceToNow` from `date-fns` here? See [docs](https://date-fns.org/v4.1.0/docs/formatDistanceToNow#). Seems like a potential simplification

ivan-ottinger replied:
Thank you, Fredrik! Yeah, I think this is something we could consider in the future. I am not that familiar with `date-fns` yet, but looks like we could definitely make some improvements and simplifications in the way we calculate relative time in Studio.

Since we decided on a more straightforward solution to address the issue at hand, I did not explore `formatDistanceToNow` further today. I have created a dedicated task: https://github.com/Automattic/dotcom-forge/issues/10449 though.

bcotrim reviewed (commented):
The code looks good and works as expected—great job!
My only suggestion is about the format. Displaying only the most significant value (e.g., `6 days` or `20 hours` if it's under a day) in the UI and adding a tooltip for the full value could make it cleaner.
What do you think about this approach?

sejas commented:
I like @bcotrim's suggestion. I think displaying the most significant value makes sense, and the tooltip is also a good suggestion.
Maybe inside the tooltip we could display the final date. Something like `Feb 12 2025 10:47:33` or any similar format.

What do you think @ivan-ottinger ?

wojtekn commented:
The short format goes well in English, but I'm unsure if everyone will well understand it. Also, it may not be that clear in other languages. This looked clearer in previous UI:

![Screenshot 2025-02-12 at 12 00 23](https://github.com/user-attachments/assets/83d448ec-bade-448a-a7b2-063eb8063574)

wojtekn commented:
What if we increased those columns' width, especially since they are too wide to fit heading text in some languages? And in worst case, added a deliberate line break after days, to avoid having "23" in one line and "hours" in another one?

ivan-ottinger commented:
> I like @bcotrim's suggestion. I think displaying the most significant value makes sense, and the tooltip is also a good suggestion.
> Maybe inside the tooltip we could display the final date. Something like Feb 12 2025 10:47:33 or any similar format.

> What do you think @ivan-ottinger ?

I have been experimenting with this in the past hour and it could be something like this for instance (I have also added word `in` there, e.g. `in 5 days`):

![Markup on 2025-02-12 at 12:56:31](https://github.com/user-attachments/assets/946ac340-9d29-4420-b2c2-2cde70c349f2)

But we could also use a full date in the tooltip as you are suggesting. 🙂 

> What if we increased those columns' width, especially since they are too wide to fit heading text in some languages? And in worst case, added a deliberate line break after days, to avoid having "23" in one line and "hours" in another one?

This also sounds good to me. It is probably the simplest approach. Full date could still be displayed in the tooltip if we wanted.

wojtekn commented:
@ivan-ottinger yes, displaying the full date in the tooltip makes sense in every case. I still have concerns about using h, m, d. They don't go well with other languages, e.g., "2d 3h 5s" would be "2 dni 3 godz. 5 sek." in Polish.

ivan-ottinger commented:
> @ivan-ottinger yes, displaying the full date in the tooltip makes sense in every case. I still have concerns about using h, m, d. They don't go well with other languages, e.g., "2d 3h 5s" would be "2 dni 3 godz. 5 sek." in Polish.

In that case we could use the full format in the field, e.g. `4 days, 2 hours`, maybe also including `in` as well: `in 4 days, 2 hours` and make the column wider.

Then, we could also add the full date into the tooltip.

What do you think?

wojtekn commented:
@ivan-ottinger displaying "4 days, 2 hours" in line and the exact date in tooltip makes sense. The "in" part can be handled as a part of the table heading. For example, in English, "Expires" makes sense alone, but in Polish, I would translate it as "Expires in" (e.g., "Wygasa za").

ivan-ottinger commented:
> @ivan-ottinger displaying "4 days, 2 hours" in line and the exact date in tooltip makes sense. The "in" part can be handled as a part of the table heading. For example, in English, "Expires" makes sense alone, but in Polish, I would translate it as "Expires in" (e.g., "Wygasa za").

Right! Makes sense. 👍🏼 

When it comes to the exact date, would we like to use the format suggested by Antonio in his comment above (`Feb 12 2025 10:47:33`)? We could keep it as is or we could translate it as well.

ivan-ottinger committed:
Adjust preview site expiration date formatting

ivan-ottinger committed:
Put back longer relative date for last updated time

ivan-ottinger committed:
Revert unnecessary changes

ivan-ottinger committed:
Fix unit test

ivan-ottinger committed:
Update width of columns

ivan-ottinger commented:
What about formatting like this? 🙂 

English:
![Markup on 2025-02-12 at 13:52:53](https://github.com/user-attachments/assets/2b6191ec-a951-4638-97bc-bcdd4a3c80e1)

Polish:
![Markup on 2025-02-12 at 13:51:03](https://github.com/user-attachments/assets/21154833-4f1f-4eaa-8001-b67585aaa732)

sejas commented on src/hooks/use-expiration-date.ts:
```diff
@@ -47,6 +63,7 @@ export function useExpirationDate( snapshotDate: number ) {
 	return {
 		isExpired,
 		countDown: isExpired ? __( 'Expired' ) : countDown,
+		fullDateString: formatStringDate( endDate.getTime(), locale, 'long' ),
```
Should we call it `expireDateString` or `fullExpireDateString` or `endDateString` ?

```suggestion
		expireDateString: formatStringDate( endDate.getTime(), locale, 'long' ),
```

ivan-ottinger replied:
That's a valid point. I named the string based on the existing `dateString` we are already returning by `useExpirationDate` when used in the current Demo sites UI.

But since we are going to remove that UI soon, I think it makes sense to improve the naming of `fullDateString`. 

`expireDateString` sounds the most clear to me.

ivan-ottinger replied:
→ Updated in https://github.com/Automattic/studio/pull/919/commits/c85f3d276bc55e8b6606ad0779532e98ba5599db.

sejas commented on src/hooks/use-expiration-date.ts:
```diff
@@ -5,12 +5,28 @@ import { useI18nData } from 'src/hooks/use-i18n-data';
+					second: '2-digit',
+					hour12: false,
+			  };
+	const formatter = new Intl.DateTimeFormat( locale, options );
```
Could we define an object with the formats instead of using a ternary operator?

Something like:
```
type FormatKey = 'short' | 'long';
```

```suggestion
	const options: Record< FormatKey, Intl.DateTimeFormatOptions > = {
		short: {
			day: 'numeric',
			month: 'long',
			year: 'numeric',
		},
		long: {
			day: '2-digit',
			month: 'short',
			year: 'numeric',
			hour: '2-digit',
			minute: '2-digit',
			second: '2-digit',
			hour12: false,
		},
	};
	const formatter = new Intl.DateTimeFormat( locale, options[ format ] );
```

ivan-ottinger replied:
That's a great idea! Thanks! I have made the change in https://github.com/Automattic/studio/pull/919/commits/e8711e3e4098243d55cbf4957fcca74fd9d40f58.

bcotrim reviewed (commented):
Maybe it's unrelated to this PR but the table behavior looks odd on smaller windows (notice the alignment of the `X hours..` and the alignment of the titles compared with the values)
![image](https://github.com/user-attachments/assets/dd640dd7-5af4-4b7f-a99d-94df30a66177)

sejas commented on src/hooks/use-expiration-date.ts:
```diff
@@ -5,12 +5,28 @@ import { useI18nData } from 'src/hooks/use-i18n-data';
+					year: 'numeric',
+					hour: '2-digit',
+					minute: '2-digit',
+					second: '2-digit',
```
I think we can remove the seconds from the long date.

```suggestion
```

sejas commented on src/hooks/use-expiration-date.ts:
```diff
@@ -5,12 +5,28 @@ import { useI18nData } from 'src/hooks/use-i18n-data';
+	locale: SupportedLocale,
+	format: 'long' | 'short' = 'short'
+): string {
+	const options: Intl.DateTimeFormatOptions =
```
Another alternative could be using the format function from `date-fns`. Something like: `format( ms, 'PPpp', { locale: dateFnsLocales[ locale ] } )`

`Intl.DateTimeFormatOptions` is also good!, and consistent with what we had before.

wojtekn replied:
Is there anything against using the solution we followed in previous UI?

sejas replied:
If you mean using `Intl.DateTimeFormatOptions`, there is nothing bad about using it, let's keep it.

sejas reviewed (changes requested):
Making the columns wider is a good solution.
Would be possible to make them more flexible, even if it means two rows?

In my case, because the domain is longer, it moves the other two columns to the right and the action menu disappears out of window producing a scroll.

<img width="1012" alt="Screenshot 2025-02-12 at 14 06 10" src="https://github.com/user-attachments/assets/05f34f63-68e1-4dcf-9fa8-0f7137342af4" />

As an alternative, I was thinking about truncate the dates 🤔 .

sejas commented:
> Maybe it's unrelated to this PR but the table behavior looks odd on smaller windows (notice the alignment of the X hours.. and the alignment of the titles compared with the values)

Good catch.
~~What if we hide the updated column when the sidebar is hidden?~~

wojtekn commented:
@sejas shouldn't the main content width match the window width of Studio that has sidebar closed?

ivan-ottinger committed:
Remove seconds from the date format

Co-authored-by: Antonio Sejas <antonio.sejas@automattic.com>

ivan-ottinger committed:
Refactor date format options

sejas commented:
> @sejas shouldn't the main content width match the window width of Studio that has sidebar closed?

You are right!, Let's not focus on the collapsed version as we can fix it by keeping the same width of the main content.

ivan-ottinger committed:
Rename `fullDateString` to `expireDateString`

ivan-ottinger committed:
Update widths of columns

ivan-ottinger commented:
Thank you everyone for your feedback! It has been helpful.

I have addressed your suggestions and adjusted the column widths a little bit (focusing only on the view when the sidebar is not collapsed).

I have tested the view with multiple different languages and different sizes of sites and haven't found other issues there, but feel free to take a look as well.

ivan-ottinger commented:
Ok, I noticed the issue with long URLs is still there when it is bit longer than what I tested with. Going to look into it further:

> In my case, because the domain is longer, it moves the other two columns to the right and the action menu disappears out of window producing a scroll.

ivan-ottinger committed:
Set max-width for the Preview site column

ivan-ottinger commented:
> Ok, I noticed the issue with long URLs is still there when it is bit longer than what I tested with. Going to look into it further:

I see we had truncate styling already applied to the first column, but it never applied since the max-width was set to 100%. I have now made the change for a fix width of the first column as well and truncating will now do its job:

![Markup on 2025-02-12 at 16:51:53](https://github.com/user-attachments/assets/603d1aa8-43df-4429-adaa-8b1ce430747d)

sejas reviewed (approved):
Looks good! The dates are displayed in a single line, and we've added a tooltip in the expires column to display the exact date.
We can improve the spacing between the columns to be consistent in a different PR.

<img width="1032" alt="Screenshot 2025-02-12 at 17 31 39" src="https://github.com/user-attachments/assets/bebcdf6d-c8a1-45ec-8f49-19cfbd8794c4" />

<img width="1032" alt="Screenshot 2025-02-12 at 17 37 04" src="https://github.com/user-attachments/assets/0daaf346-f1d8-4def-88a1-71e8b779314c" />
